service/history/api/multioperation/api.go

```diff
@@ -73,146 +87,77 @@ func Invoke(
+		resp, err := mo.startAndUpdateWorkflow(ctx)
 		var noStartErr *noStartError
 		switch {
 		case errors.As(err, &noStartErr):
-			// The start request was deduped. Continue below and only send the update.
+			// The start request was deduped, no termination is needed.
+			// Continue below by only sending the update.
 		case err != nil:
 			return nil, err
 		default:
 			return resp, nil
 		}
 	}
 
-	currentWorkflowLease, err := workflowConsistencyChecker.GetWorkflowLease(
-		ctx,
-		nil,
-		definition.NewWorkflowKey(req.NamespaceId, startReq.StartRequest.WorkflowId, ""),
-		locks.PriorityHigh,
-	)
-	var notFound *serviceerror.NotFound
-	if errors.As(err, &notFound) {
-		currentWorkflowLease = nil
-	} else if err != nil {
-		return nil, newMultiOpError(err, multiOpAbortedErr)
+	runningWorkflowLease, err := mo.getRunningWorkflowLease(ctx)
+	if err != nil {
+		return nil, err
 	}
 
-	// workflow was already started, ...
-	if currentWorkflowLease != nil {
-		if !currentWorkflowLease.GetMutableState().IsWorkflowExecutionRunning() {
-			// ... but cannot receive an update since it's not running.
-			currentWorkflowLease.GetReleaseFn()(nil)
-		} else {
-			// and is running, ...
-			switch conflictPolicy {
-			case enumspb.WORKFLOW_ID_CONFLICT_POLICY_USE_EXISTING:
-				// ... skip the start and only send the update
-				// NOTE: currentWorkflowLease will be released by the function
-				return updateWorkflow(ctx, shardContext, currentWorkflowLease, updater)
-
-			case enumspb.WORKFLOW_ID_CONFLICT_POLICY_FAIL:
-				// ... if same request ID, just send update
-				// NOTE: currentWorkflowLease will be released by the function
-				if dedup(startReq, currentWorkflowLease) {
-					return updateWorkflow(ctx, shardContext, currentWorkflowLease, updater)
-				}
-
-				// ... otherwise, abort the entire operation
-				currentWorkflowLease.GetReleaseFn()(nil) // nil since nothing was modified
-				wfKey := currentWorkflowLease.GetContext().GetWorkflowKey()
-				err = serviceerror.NewWorkflowExecutionAlreadyStarted(
-					fmt.Sprintf("Workflow execution is already running. WorkflowId: %v, RunId: %v.", wfKey.WorkflowID, wfKey.RunID),
-					startReq.StartRequest.RequestId,
-					wfKey.RunID,
-				)
-				return nil, newMultiOpError(err, multiOpAbortedErr)
-
-			case enumspb.WORKFLOW_ID_CONFLICT_POLICY_TERMINATE_EXISTING:
-				return updateWorkflow(ctx, shardContext, currentWorkflowLease, updater)
-
-			case enumspb.WORKFLOW_ID_CONFLICT_POLICY_UNSPECIFIED:
-				// ... fail since this policy is invalid
-				currentWorkflowLease.GetReleaseFn()(nil) // nil since nothing was modified
-				return nil, serviceerror.NewInvalidArgument("unhandled workflow id conflict policy: unspecified")
-			}
+	// Workflow was already started ...
+	if runningWorkflowLease != nil {
+		if err = mo.allowUpdateWorkflow(ctx, runningWorkflowLease, conflictPolicy); err != nil {
+			runningWorkflowLease.GetReleaseFn()(nil) // nil since nothing was modified
```

Comment:
Moved the release here; and made `allowUpdateWorkflow` free of locking concerns.